### PR TITLE
Use LocationDataSource

### PR DESCRIPTION
**Pull request by mhdostal**
Update ArcGISARView to use the newly updated LocationDataSource class and subclasses.  Some updates for latest design.

### CODE REVIEW

File: Examples/ArcGISToolkitExamples/ARExample.swift [hunk 1]
```diff
@@ -36,10 +36,13 @@ class ARExample: UIViewController {
             ])
         
         arView.sceneView.scene = makeStreetsScene()
+        arView.locationDataSource = AGSCLLocationDataSource()
     }
     
     override func viewDidAppear(_ animated: Bool) {
```
Comment (philium) [inline Examples/ArcGISToolkitExamples/ARExample.swift]: I noticed that you're not calling `super`:

> You can override this method to perform additional tasks associated with presenting the view. If you override this method, you must call `super` at some point in your implementation.

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 2]
```diff
@@ -16,104 +16,86 @@ import UIKit
...
-    public var arConfiguration: ARConfiguration = {
-        let config = ARWorldTrackingConfiguration()
-        config.worldAlignment = .gravityAndHeading
-        return config
-        }() {
+    /// Denotes whether ARKit is being used to track location and angles.
+    public private(set) var isUsingARKit: Bool = true
+
+    /// The data source used to get device location.  Used either in conjuction with ARKit data or when ARKit is not present or not being used.
+    public var locationDataSource: AGSCLLocationDataSource? {
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]: I just want to confirm that this is intended to be publicly settable?

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 3]
```diff
@@ -166,71 +165,73 @@ public class ArcGISARView: UIView {
...
-        }
-        
-        if originCamera != nil {
-            // We have a starting camera, so no need to start the location manager, just finalizeStart().
-            finalizeStart()
+    public func startTracking(_ completion: ((_ error: Error?) -> Void)? = nil) {
+        // We have a location data source that needs to be started.
+        if let locationDataSource = self.locationDataSource {
+            locationDataSource.start { [weak self] (error) in
+                if let error = error {
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]: I think you could simplify this slightly by only calling `completion` in one place:

```swift
if error == nil {
    self?.finalizeStart()
}
completion?(error)
```

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 4]
```diff
@@ -166,71 +165,73 @@ public class ArcGISARView: UIView {
...
-        arSCNView.isHidden = !renderVideoFeed
-        
-        // Run the ARSession.
-        arSCNView.session.run(arConfiguration, options:.resetTracking)
-        didStartOrFailWithError(nil)
+        DispatchQueue.main.async { [weak self] in
+            guard let strongSelf = self else { return }
+            // Run the ARSession.
+            if strongSelf.isUsingARKit {
+                strongSelf.arSCNView.session.run(strongSelf.arConfiguration, options:.resetTracking)
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]:
```suggestion
                strongSelf.arSCNView.session.run(strongSelf.arConfiguration, options: .resetTracking)
```

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 5]
```diff
@@ -400,9 +335,9 @@ extension ArcGISARView: SCNSceneRendererDelegate {
                                                            quaternionY: finalQuat.vector.y,
                                                            quaternionZ: finalQuat.vector.z,
                                                            quaternionW: finalQuat.vector.w,
-                                                           translationX: (cameraTransform.columns.3.x) * translationTransformationFactor,
-                                                           translationY: (-cameraTransform.columns.3.z) * translationTransformationFactor,
-                                                           translationZ: (cameraTransform.columns.3.y) * translationTransformationFactor)
+                                                           translationX: (cameraTransform.columns.3.x),
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]:
```suggestion
                                                           translationX: cameraTransform.columns.3.x,
```

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 6]
```diff
@@ -400,9 +335,9 @@ extension ArcGISARView: SCNSceneRendererDelegate {
...
+                                                           translationY: (-cameraTransform.columns.3.z),
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]:
```suggestion
                                                           translationY: -cameraTransform.columns.3.z,
```

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift [hunk 7]
```diff
@@ -400,9 +335,9 @@ extension ArcGISARView: SCNSceneRendererDelegate {
...
+                                                           translationZ: (cameraTransform.columns.3.y))
```
Comment (philium) [inline Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]:
```suggestion
                                                           translationZ: cameraTransform.columns.3.y)
```

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift (hunk 2 above)
Comment (mhdostal) [reply Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]: Yes, publicly settable.  We allow clients to create and use their own LocationDataSources, like from a high-accuracy GPS or maybe Beacons.

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift (hunk 3 above)
Comment (mhdostal) [reply Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]: Perfect.

File: Toolkit/ArcGISToolkit/AR/ArcGISARView.swift (hunk 4 above)
Comment (mhdostal) [reply Toolkit/ArcGISToolkit/AR/ArcGISARView.swift]: Done.

File: Examples/ArcGISToolkitExamples/ARExample.swift [hunk 8]
```diff
@@ -36,10 +36,13 @@ class ARExample: UIViewController {
...
-        arView.startTracking()
+        arView.startTracking { (error) in
+            print("Error starting ArcGISARView tracking: \(String(describing: error))")
```
[the remark this answers is not in the document]
Comment (mhdostal) [reply Examples/ArcGISToolkitExamples/ARExample.swift]: Done.

**Comment by mhdostal** [comment]
@pgruenler I made the changes.

File: Examples/ArcGISToolkitExamples/ARExample.swift [hunk 9]
```diff
@@ -36,13 +36,20 @@ class ARExample: UIViewController {
...
         
         arView.sceneView.scene = makeStreetsScene()
+        arView.locationDataSource = AGSCLLocationDataSource()
     }
     
     override func viewDidAppear(_ animated: Bool) {
-        arView.startTracking()
+        super.viewDidAppear(animated)
+        arView.startTracking { (error) in
+            if error != nil {
```
Comment (philium) [inline Examples/ArcGISToolkitExamples/ARExample.swift]: I would suggesting binding the error:

```swift
if let error = error {
    print("Error starting ArcGISARView tracking: \(error)")
}
```

**Pull request merged**